Guides/Keyed.md

```diff
@@ -0,0 +1,74 @@
+# Indexed
+
+[[Source](https://github.com/apple/swift-algorithms/blob/main/Sources/Algorithms/Indexed.swift) | 
+ [Tests](https://github.com/apple/swift-algorithms/blob/main/Tests/SwiftAlgorithmsTests/IndexedTests.swift)]
+
+Stores the elements of a sequence as the values of a Dictionary, keyed by the result of the given closure.
+
+```swift
+let fruits = ["Apple", "Banana", "Cherry"]
+let fruitByLetter = fruits.keyed(by: { $0.first! })
+// Results in:
+[
+	"A": "Apple",
+	"B": "Banana",
+	"C": "Cherry",
+]
+```
+
+Duplicate keys will trigger a runtime error by default. To handle this, you can provide a closure which specifies which value to keep:
```

Comment:
This makes the method a bit dangerous, since a naive user could fail to realise keys might collide.  Even a knowledgeable user might make the simple mistake of forgetting to provide the `uniquingKeysWith` argument (e.g. auto-complete omits it and they don't notice in the moment).

It makes it more verbose to handle 'unsafe' data - arguably the common case in today's network-centric world - since you have to implement the second closure (as merely a throw statement) in order to avoid crashing on bad input.

These problems can be addressed by having two versions of the method:

```swift
public func keyed<Key>(
    by keyForValue: (Element) throws -> Key
) throws -> [Key: Element]

public func keyed<Key>(
    by keyForValue: (Element) throws -> Key,
    uniquingKeysWith combine: (Key, Element, Element) throws -> Element
) rethrows -> [Key: Element]
```

This preserves the ability to have a non-throwing version - and thus not require use of `try` - iff the second form is used and both closures don't throw, which is the only case where it's compile-time assured that it cannot fail.

Any user can still use `try!` to mimic the other behaviour (of crashing), if they wish.  And the explicit presence of that `!` in their code is an important warning that it is potentially unsafe.

(it also has the minor benefit of not allowing you to explicitly set `uniquingKeysWith` to nil, which has ambiguous and potentially misleading intent, e.g. a reader might assume that makes it not unique keys (somehow))

Reply:
> These problems can be addressed by having two versions of the method:

I love this idea! I'll make these changes

> (it also has the minor benefit of not allowing you to explicitly set uniquingKeysWith to nil, which has ambiguous and potentially misleading intent, e.g. a reader might assume that makes it not unique keys (somehow))

Good call!